[Docs] Add V1.4.0 to change log (#3907)
mlrun/mlrun: docs/change-log/index.md

```
 # Set 2 functions:
func1 = project.set_function(...)
func2 = project.set_function(...)

# Run the first function:
run1 = func1.run(...)
# In the function  `func1` we logged a model "my_model" and an artifact "my_artifact"
run1.outputs  
{
    "my_model": "store://...",
    "my_artifact": "store://...",  # Instead of target path: "/User/.../data.csv"
}

# The function `func2` expects a `DataItem` for the logged artifact so passing it through inputs will work as `DataItem` can work with store paths:
run2 = func2.run(..., inputs={"artifact": run1.outputs["my_artifact"]})

# But passing it through a parameter won't work as the string value is now a store path and not a target path:
run2 = func2.run(..., params={"artifact": run1.outputs["my_artifact"]})
```

### Closed issues

| ID   | Description                                                    |
| --- | ----------------------------------------------------------------- |

> | ML-1787 | Optimized distribution of load between chief and workers so that heavy loads do not cause restart of kubelet. [View in Git](https://github.com/mlrun/mlrun/pull/). |

Author: I am reviewing again now and cleaning up these loose ends

The MLRun server is now based on Python 3.9. It's recommended to move the client to Python 3.9 as well. 

MLRun v1.3.0 maintains support for mlrun base images that are based on python 3.7. To differentiate between the images, the images based on
python 3.7 have the suffix: `-py37`. The correct version is automatically chosen for the built-in MLRun images according to the Python version of the MLRun client (for example, a 3.7 Jupyter gets the `-py37` images).

MLRun v1.3.x maintains support for mlrun base images that are based on a python 3.7 environment. To differentiate between the images, the images based on
python 3.7 have the suffix: `-py37`. The correct version is automatically chosen for the built-in MLRun images according to the Python version of the MLRun client (for example, a 3.7 Jupyter gets the `-py37` images).

> For a Python 3.9 environment see [Set up a Python 3.7 client environment](../install/remote.html#set-up-a-python-3-9-client-environment).

Reviewer: ```suggestion
For a Python 3.9 environment see [Set up a Python 3.9 client environment](../install/remote.html#set-up-a-python-3-9-client-environment).
```

| ML-3381 | Private repo is not supported as a marketplace hub             | NA                                     | v1.2.1 | 
| ML-3520 | MLRun does not decompress large Kubeflow pipelines | NA | v1.3.0 |
| ML-3824 | MLRun supports TensorFlow up to 2.11. | NA | v1.3.1 |
| ML-3731 | When trying to identify a failed step in a workflow with `mlrun.get_run_db().list_pipelines('project-name')`, the returned error is `None`. | To see the error, use `mlrun.db.get_pipelines()` instead. |
| ML-3743 | Setting AWS credentials as project secret cause a build failure on EKS configured with ECR. | When using an ECR as the external container registry, make sure that the project secrets AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY have read/write access to ECR, 
as described in the [platform documentation](https://www.iguazio.com/docs/latest-release/services/app-services/docker-registry/#create-off-cluster-registry) 

## Deprecations

    

> | In   | ID |Description                                  ]]                        |

Reviewer: is the `]]` on purpose?

Author: deleted

| `mount_v3io_extended`                            | `mount_v3io`                   |
| `LegacyArtifact` and all legacy artifact types that inherit from it (`LegacyArtifact`, `LegacyDirArtifact`, `LegacyLinkArtifact`, `LegacyPlotArtifact`, `LegacyChartArtifact`, `LegacyTableArtifact`, `LegacyModelArtifact`, `LegacyDatasetArtifact`, `LegacyPlotlyArtifact`, `LegacyBokehArtifact`) | `Artifact` or other artifact classes that inherit from it |
| `init_functions` in pipelines                    | Add the function initialization to the pipeline code instead |
| The entire `mlrun/mlutils` library               | `mlrun.framework`                     |
| `run_pipeline`                                   | `project.run`                                     |

### Deprecated APIs, will be removed in v1.6.0
These APIs will be removed from the v1.6.0 code. A FutureWarning appears if you try to use them in v1.6.0.
| Deprecated / to be removed                       | Use instead                                   |
| ------------------------------------------------ | --------------------------------------------- |
| clear_context() |  |

> ### Deprecated CLI, will be removed in v1.5.0

Reviewer: As I see it this section should be above the "will be removed in v1.6.0" one

Author: I grouped them by API/CLI and not release. Both work - but switched them

## Future deprecations

| ID   | When | Description                                                    |
| --- | ----| ----------------------------------------------------------------- |
| ML-3605 | v1.5.0 | Model Monitoring:  Most of the charts and KPIs in Grafana are now based on the data store target instead of the MLRun API. It is recommended to update the model monitoring dashboards since the old dashboards won't be supported. |

## Deprecated APIs and CLI

### APIs deprecated in v1.4.0, will be removed from v1.6.0 code
These APIs will be removed from the v1.6.0 code. A FutureWarning appears if you try to use them in v1.4.0 and higher.

| Deprecated / to be removed                       | Use instead                                   |
| ------------------------------------------------ | --------------------------------------------- |
| MLRunProjectclear_context() | This method deletes all files and clears the context directory or subpath (if defined). This method can produce unexpected outcomes and is not recommended.  |
| MLRunProject object legacy parameters | metadata and spec instead |

> | BaseRuntime.with_commands and KubejobRuntime.build_config'verify_base_image' param | 'prepare_image_for_deploy' |

Reviewer: ```suggestion
| BaseRuntime.with_commands and KubejobRuntime.build_config 'verify_base_image' param | 'prepare_image_for_deploy' |
```

| run_local | function.run(local=True) |

### REST APIs deprecated in v1.4.0, will be removed from v1.6.0 code
| Deprecated                        | Use instead                                   |
| ------------------------------------------------ | --------------------------------------------- |

> | http request POST /artifact/{project}/{uid}/{key:path} | /projects/{project}/artifacts/{uid}/{key:path} instead |

Reviewer: Saying "REST APIs"  in the title is enough :)
```suggestion
| POST /artifact/{project}/{uid}/{key:path} | /projects/{project}/artifacts/{uid}/{key:path} instead |
```

| v1.4.0 |  ML-3547 | [APIs deprecated in v1.4.0, will be removed from v1.6.0 code](#apis-deprecated-in-v1-4-0-will-be-removed-from-v1-6-0-code) and [REST APIs deprecated in v1.4.0, will be removed from v1.6.0 code](#rest-apis-deprecated-in-v1-4-0-will-be-removed-from-v1-6-0-code).|

## Future deprecations

| ID   | When | Description                                                    |
| --- | ----| ----------------------------------------------------------------- |
| ML-3605 | v1.5.0 | Model Monitoring:  Most of the charts and KPIs in Grafana are now based on the data store target instead of the MLRun API. It is recommended to update the model monitoring dashboards since the old dashboards won't be supported. |

## Deprecated APIs and CLI

### APIs deprecated in v1.4.0, will be removed from v1.6.0 code
These APIs will be removed from the v1.6.0 code. A FutureWarning appears if you try to use them in v1.4.0 and higher.

| Deprecated / to be removed                       | Use instead                                   |
| ------------------------------------------------ | --------------------------------------------- |

> | MLRunProjectclear_context() | This method deletes all files and clears the context directory or subpath (if defined). This method can produce unexpected outcomes and is not recommended.  |

Reviewer: ```suggestion
| MLRunProject.clear_context() | This method deletes all files and clears the context directory or subpath (if defined). This method can produce unexpected outcomes and is not recommended.  |
```

| MLRunProject object legacy parameters | metadata and spec instead |
| BaseRuntime.with_commands and KubejobRuntime.build_config 'verify_base_image' param | 'prepare_image_for_deploy' |
| run_local | function.run(local=True) |

### REST APIs deprecated in v1.4.0, will be removed from v1.6.0 code
| Deprecated                        | Use instead                                   |
| ------------------------------------------------ | --------------------------------------------- |
| POST /artifact/{project}/{uid}/{key:path} | /projects/{project}/artifacts/{uid}/{key:path} instead |

> | http request GET /projects/{project}/artifact/{key:path} | /projects/{project}/artifacts/{key:path} instead |
> | http request DELETE /artifact/{project}/{uid} | /projects/{project}/artifacts/{uid} instead |
> | http request GET /artifacts | /projects/{project}/artifacts instead |
> | http request DELETE /artifacts  | /projects/{project}/artifacts instead |
> | http request POST /func/{project}/{name}  | /projects/{project}/functions/{name} instead |
> | http request GET /func/{project}/{name}  | /projects/{project}/functions/{name} instead |
> | http request GET /funcs  | /projects/{project}/functions instead |

Reviewer: ```suggestion
| GET /projects/{project}/artifact/{key:path} | /projects/{project}/artifacts/{key:path} instead |
| DELETE /artifact/{project}/{uid} | /projects/{project}/artifacts/{uid} instead |
| GET /artifacts | /projects/{project}/artifacts instead |
| DELETE /artifacts  | /projects/{project}/artifacts instead |
| POST /func/{project}/{name}  | /projects/{project}/functions/{name} instead |
| GET /func/{project}/{name}  | /projects/{project}/functions/{name} instead |
| GET /funcs  | /projects/{project}/functions instead |
```
